webapp/tests/karma/ts/services/rules-engine.service.spec.ts

```diff
@@ -0,0 +1,619 @@
+      expect(rulesEngineCoreStubs.updateEmissionsFor.callCount).to.equal(0);
+    });
+
+    it('should only filter tasks and refresh requesters', async () => {
+      service = TestBed.inject(RulesEngineService);
+      const replicationResult = {
+        doc_write_failures: 0,
+        docs_read: 6,
+        docs_written: 6,
+        errors: [],
+        ok: true,
+        last_seq: 20,
+        docs: [
+          { _id: 'report1', fields: { patient_id: 'patient' }, type: 'data_record' },
+          { _id: 'task~1', emission: { _id: '??' }, requester: 'patient_uuid', type: 'task' },
+          { _id: 'contact2', type: 'contact', name: 'C', patient_id: 'patient2' },
+          { _id: 'task~2', emission: { _id: '??' }, requester: 'other_patient', type: 'task' },
+          { _id: 'task~3', emission: { _id: '!!' }, requester: 'other_patient', type: 'task' },
+          { _id: 'report2', fields: { patient_uuid: 'etc' }, type: 'data_record' },
+        ]
+      };
+
+      await service.monitorExternalChanges(replicationResult);
+
+      expect(rulesEngineCoreStubs.updateEmissionsFor.args[0]).to.deep.equal([['patient_uuid', 'other_patient']]);
+      expect(rulesEngineCoreStubs.updateEmissionsFor.callCount).to.equal(1);
+    });
+  });
+
+  it('fetchTaskDocsForAllContacts() should fetch task for all contacts', async () => {
+    const taskDoc = JSON.parse(JSON.stringify(sampleTaskDoc));
+    fetchTasksResult = sinon.stub().resolves([taskDoc]);
+    rulesEngineCoreStubs.getDirtyContacts.returns(['a', 'b', 'c']);
+    service = TestBed.inject(RulesEngineService);
+
+    const actual = await service.fetchTaskDocsForAllContacts();
+
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTasksFor.args[0][0]).to.be.undefined;
+    expect(actual.length).to.eq(1);
+    expect(actual[0]).to.nested.include({
+      _id: 'taskdoc',
+      'emission.other': true,
+    });
+    expect(telemetryService.record.callCount).to.equal(4);
+    expect(telemetryService.record.args[1]).to.deep.equal(['rules-engine:tasks:dirty-contacts', 3]);
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:ensureTaskFreshness:cancel');
+    expect(telemetryService.record.args[3][0]).to.equal('rules-engine:tasks:all-contacts');
+  });
+
+  it('fetchTaskDocsFor() should fetch task docs', async () => {
+    const taskDoc = JSON.parse(JSON.stringify(sampleTaskDoc));
+    const contactIds = ['a', 'b', 'c'];
+    fetchTasksResult = sinon.stub().resolves([taskDoc]);
+    rulesEngineCoreStubs.getDirtyContacts.returns(['a', 'b']);
+    service = TestBed.inject(RulesEngineService);
+
+    const actual = await service.fetchTaskDocsFor(contactIds);
+
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTasksFor.args[0][0]).to.eq(contactIds);
+    expect(actual.length).to.eq(1);
+    expect(actual[0]).to.nested.include({
+      _id: 'taskdoc',
+      'emission.other': true,
```

Comment:
it looks like you're ommiting testing that some fields from these tasks are translated by the app.

Reply:
I'm going to see if I can mock the translateService.instant() for test.

Reply:
Done